ipc/provider/src/manager/evm/dryrun.rs

```diff
@@ -0,0 +1,405 @@
+
+use crate::config::serialize::{serialize_address_to_str, serialize_bytes_to_str};
+use crate::manager::{
+    BottomUpCheckpointRelayer, EthSubnetManager, GetBlockHashResult, SubnetGenesisInfo, SubnetInfo,
+    SubnetManager, TopDownFinalityQuery, TopDownQueryPayload,
+};
+use crate::VMType;
+use fvm_ipld_encoding::{BytesSer, RawBytes};
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::MethodNum;
+use ipc_api::checkpoint::{
+    BottomUpCheckpoint, BottomUpCheckpointBundle, QuorumReachedEvent, Signature,
+};
+use ipc_api::cross::IpcEnvelope;
+use ipc_api::eth_to_fil_amount;
+use ipc_api::ethers_address_to_fil_address;
+use ipc_api::staking::{StakingChangeRequest, ValidatorInfo};
+use ipc_api::subnet::ConstructParams;
+use ipc_api::subnet_id::SubnetID;
+use serde::Serialize;
+
+const INVOKE_CONTRACT: MethodNum = 3844450837;
+
+#[derive(Serialize)]
+struct MockedFVMTxn {
+    #[serde(serialize_with = "serialize_address_to_str")]
+    from: Address,
+    #[serde(serialize_with = "serialize_address_to_str")]
+    to: Address,
+    /// The value, display as string to align with evm display
+    value: String,
+    #[serde(serialize_with = "serialize_bytes_to_str")]
+    method_params: Vec<u8>,
+    method: MethodNum,
+}
+
+#[derive(Serialize)]
+struct MockedEVMTxn {
+    from: ethers::types::Address,
+    to: ethers::types::Address,
+    value: U256,
+    #[serde(serialize_with = "serialize_bytes_to_str")]
+    calldata: Vec<u8>,
+    #[serde(serialize_with = "serialize_bytes_to_str")]
+    method: Selector,
+}
+
+impl TryFrom<MockedEVMTxn> for MockedFVMTxn {
+    type Error = anyhow::Error;
+
+    fn try_from(value: MockedEVMTxn) -> Result<Self, Self::Error> {
+        Ok(Self {
+            from: ethers_address_to_fil_address(&value.from)?,
+            to: ethers_address_to_fil_address(&value.to)?,
+            value: eth_to_fil_amount(&value.value)?.to_string(),
+            method_params: to_fvm_calldata(&value.calldata)?,
+            method: INVOKE_CONTRACT,
+        })
+    }
+}
+
+/// EVM subnet dry run handler. For call transactions, it will directly execute the transaction and
+/// send to the chain to fetch actual data. For send transactions, it prints the payload to console.
+pub struct EvmSubnetDryRun {
+    /// The vm type that determines the message format
+    vm_type: VMType,
+    /// The subnet manager caller
+    caller: EthSubnetManager,
+}
+
+#[async_trait]
+impl SubnetManager for EvmSubnetDryRun {
+    async fn create_subnet(
+        &self,
+        from: Address,
+        params: ConstructParams,
+    ) -> anyhow::Result<Address> {
+        let evm = self.create_subnet_evm(&from, params)?;
+
+        let s = match self.vm_type {
+            VMType::Evm => serde_json::to_string(&evm)?,
+            VMType::Fvm => serde_json::to_string(&MockedFVMTxn::try_from(evm)?)?,
+        };
+
+        println!("txn to send: \n {}", s);
```

Comment:
If we want to pipe the output to `jq` then it would be better not to mix in non-JSON lines into the output. If you must, you can use logging (which goes to stderr), or maybe wrap the JSON into another one like `{"create_subnet": s}` in case you intend to print multiple outputs (because your dry-run doesn't stop at the first but keeps going with fake data).

Reply:
When I tried it, actually we dont really need this line. I will remove it.